features/upgrade/sdn/policy-upgrade.feature

```diff
@@ -78,3 +78,116 @@ Feature: SDN compoment upgrade testing
+    Given 2 pods become ready with labels:
+      | name=test-pods |
+    And evaluation of `pod(4).name` is stored in the :pod5 clipboard
+
+    Given I obtain test data file "rc/idle-rc-1.yaml"
+    And I run the :create client command with:
+      | f | idle-rc-1.yaml |
+    Then the step should succeed
+    Given 2 pods become ready with labels:
+      | name=hello-idle |
+    And evaluation of `pod(6).name` is stored in the :pod7 clipboard
+
+    Given I have a project
+    When I run the :label admin command with:
+      | resource | namespace                |
+      | name     | test2                    |
+      | key_val  | team=operations          |
+    Then the step should succeed
+
+    Given the AllowNamespaceAndPod policy is applied to the "test1" namespace
+    Then the step should succeed
+
+    When I use the "test1" project
+    When I execute on the "<%= cb.pod1 %>" pod:
+      | curl | -s | --connect-timeout | 5 | <%= cb.pod2ip %>:8080 |
+    Then the step should fail 
+    And the output should not contain "Hello"
+    When I execute on the "<%= cb.pod3 %>" pod:
+      | curl | -s | --connect-timeout | 5 | <%= cb.pod2ip %>:8080 |
+    Then the step should fail
+    And the output should not contain "Hello"
+
+    When I use the "test2" project
+    When I execute on the "<%= cb.pod5 %>" pod:
+      | curl | -s | --connect-timeout | 5 | <%= cb.pod2ip %>:8080 |
+    Then the step should succeed
+    And the output should contain "Hello"
+    When I execute on the "<%= cb.pod7 %>" pod:
+      | curl | -s | --connect-timeout | 5 | <%= cb.pod2ip %>:8080 |
+    Then the step should fail 
+    And the output should not contain "Hello"
+
+  # @author asood@redhat.com
+  # @case_id OCP-38751
+  @admin
+  @upgrade-check
+  Scenario: Check the networkpolicy works well after upgrade
+    Given I switch to cluster admin pseudo user
+    When I use the "test1" project
+    Given status becomes :running of 2 pods labeled:
+      | name=test-pods |
+    And evaluation of `pod(1).ip` is stored in the :pod2ip clipboard
+    And evaluation of `pod(0).name` is stored in the :pod1 clipboard
+    When I execute on the "<%= cb.pod1 %>" pod:
+      | curl | -s | --connect-timeout | 5 | <%= cb.pod2ip %>:8080 |
+    Then the step should fail
+    Given status becomes :running of 2 pods labeled:
+      | name=hello-idle |
+    And evaluation of `pod(2).name` is stored in the :pod3 clipboard
+    When I execute on the "<%= cb.pod3 %>" pod:
+      | curl | -s | --connect-timeout | 5 | <%= cb.pod2ip %>:8080 |
+    Then the step should fail
+
+    When I use the "test2" project
+    Given status becomes :running of 2 pods labeled:
+      | name=test-pods |
+    And evaluation of `pod(4).name` is stored in the :pod5 clipboard
+    When I execute on the "<%= cb.pod5 %>" pod:
+      | curl | -s | --connect-timeout | 5 | <%= cb.pod2ip %>:8080 |
+    Then the step should succeed
+    Given status becomes :running of 2 pods labeled:
```

Comment:
Here we need to add one more steps to make sure 'pod5 cannot access pod3.ip ( since pod do not have correct label)

Reply:
pod3 has label name=hello-idle, pod5 has label name=test-pods.
So pod with label name=test-pods not being accessed by pod with label name=hello-idle is covered by the following steps:-

When I execute on the "<%= cb.pod7 %>" pod:
      | curl | -s | --connect-timeout | 5 | <%= cb.pod2ip %>:8080 |
    Then the step should fail
    And the output should not contain "Hello"
Though the pods are in different namespaces/project where pod2ip is ip of the pod with label name=test-pods and pod7 is pod with label name=hello-idle, these steps do test the flow you are referring to.

Reply:
both pod3 and pod7 do not match :
namespace with label " team=operation' and pods label 'name=test-pods'
pod3 only has pod label 'name=test-pods', however it's namespace do not have label 'team=operation'
pod7 only has namespace label 'team=operation', However without pod label 'name=test-pods'

So we need one more pod (pod5) match both label (namespace with label " team=operation' and pods label 'name=test-pods'). 

- it can access pods with label 'name=test-pods' in namespace test1  ----> the current steps already including this checkpoint

- but it cannot access pods without label 'name=test-pods' in in namespace test1  ---> **we should add this checkpoint**

Reply:
you can see the policy below
```
spec:
  podSelector:
    matchLabels:
      name: test-pods       --> here is for matching namespace with this policy .  means test1 in our case
  ingress:
    - from:
      - namespaceSelector:
          matchLabels:
            team: operations
        podSelector:
          matchLabels:
            name: test-pods                 --> here is for matching test2 namespace pods in our case
```

Reply:
Scenario is pod labeled test-pods is only accessible by the pod test-pods in namespace policy-upgrade2
Putting in more details for the cases.
Namespace policy-upgrade1
hello-idle-5dftl   1/1     Running   0          141m   10.131.0.87   (hello-idle/ pod2)   
test-rc-jgtkr      1/1     Running   0          141m   10.131.0.86     (test-pods/ pod1)    
test-rc-rvccc      1/1     Running   0          141m   10.129.2.43  (test-pods/pod2ip)
Namespace policy-upgrade2
hello-idle-7chxv   1/1     Running   0          (hello-idle/ pod5)
test-rc-zvstp      1/1     Running   0          (test-pods/pod4)

Case 1 Curl from pod1 -> pod2ip fails, this is case where both pods have same label test-pods in namespace that is not labeled (I think you are referring to this case in your comment, this is already covered) 
Case 2 Curl from pod3->pod2ip fails, this case where hello-idle is trying to get to pod labeled test-pods in namespace that is not labeled.
Case 3 Curl from pod4->pod2ip pass, this is case where test-pods in upgrade-policy2 is accesses test-pods in upgrade-policy1
Case 4 Curl from pod5->pod2ip fails this is case where hello-idle in upgrade-policy2 is trying to access test-pods in upgrade-policy1.

Reply:
Actually I want to add is :  pod4 -> pod2(hello-idle-5dftl 1/1 Running 0 141m 10.131.0.87 (hello-idle/ pod2)),  this should be failed since pod2 do not have label 'name=test-pods'

others looks good to me

Reply:
This will succeed as these pods are not selected by the podSelector specified in network policy or in other words network policy does not impact these pods.
Nework policy is about the test-pods being accessible pods that are labeled test-pods and are in namspace that is labeled team=operations.

Reply:
I saw you update `hello-idle-5dftl 1/1 Running 0 141m 10.131.0.87 (hello-idle/ pod3)` to pod3 from pod2

then pod4 -> pod3 should be failed. 
as you said, this pod did not be selected by podSelectror.  so it should be accessed from pod4.

Reply:
Yes, I fixed the pod number, it was incorrect. I will add another case, though there are bunch of others that could be added.

Reply:
Added another case.